File: flash_attn/attn_profile.py
import numpy as np
import time


def fa2_forward_profile(Q, Kt, V, T, d, M_bytes, scale):
    I = M_bytes // (4 * d * 4)
    J = min(I, d)
    if I < 1: I = 1
    if J < 1: J = 1

    Tr = (T + I - 1) // I
    Tc = (T + J - 1) // J

    O    = np.zeros((T, d), dtype=np.float32)
    perf = time.perf_counter
    t7 = t8 = t10 = t11 = 0.0

    t0 = perf()

    for i in range(Tr):
        row_start = i * I
        row_end   = min(row_start + I, T)
        cur_I     = row_end - row_start

        l = np.zeros(cur_I, dtype=np.float32)
        m = np.full(cur_I, -np.inf, dtype=np.float32)
        Q_i = Q[row_start:row_end, :]

        for j in range(Tc):
            col_start = j * J
            col_end   = min(col_start + J, T)

            # Line 7: tile loads
            tt   = perf()
            Kt_j = np.asfortranarray(Kt[:, col_start:col_end])
            V_j  = np.ascontiguousarray(V[col_start:col_end, :])
            t7  += perf() - tt

            # Line 8: S = Q_i @ Kt_j
            tt   = perf()
            S    = (Q_i @ Kt_j) * scale
            t8  += perf() - tt

            # Line 10: softmax statistics
            tt     = perf()
            m_block = S.max(axis=1)
            m_new   = np.maximum(m, m_block)
            alpha   = np.exp(m - m_new)
            P       = np.exp(S - m_new[:, None])
            l_new   = alpha * l + P.sum(axis=1)
            t10    += perf() - tt

            # Line 11: output accumulation
            tt  = perf()
            O[row_start:row_end, :] = (O[row_start:row_end, :] * alpha[:, None]) + (P @ V_j)
            t11 += perf() - tt

            m, l = m_new, l_new

        O[row_start:row_end, :] /= l[:, None]

    runtime = perf() - t0
    others  = max(0.0, runtime - (t7 + t8 + t10 + t11))
    return O, runtime, t7, t8, t10, t11, others

File: src/attn_profile.cpp
#include "flash_attn/attn.hpp"
#include <algorithm>
#include <chrono>
#include <cmath>

ProfileResult fa2_forward_profile(const MatrixXfRM& Q, const MatrixXfRM& Kt,
                                   const MatrixXfRM& V, int T, int d,
                                   int M_bytes, float scale) {
    int I = M_bytes / (4 * d * 4);
    int J = std::min(I, d);
    if (I < 1) I = 1;
    if (J < 1) J = 1;

    const int Tr = (T + I - 1) / I;
    const int Tc = (T + J - 1) / J;

    MatrixXfRM O = MatrixXfRM::Zero(T, d);
    ProfileData p{};

    using Clock = std::chrono::steady_clock;
    const auto t0 = Clock::now();

    for (int i = 0; i < Tr; ++i) {
        const int row_start = i * I;
        const int cur_I     = std::min(row_start + I, T) - row_start;

        Eigen::VectorXf l = Eigen::VectorXf::Zero(cur_I);
        Eigen::VectorXf m = Eigen::VectorXf::Constant(cur_I, -INFINITY);
        MatrixXfRM Q_i = Q.middleRows(row_start, cur_I);

        for (int j = 0; j < Tc; ++j) {
            const int col_start = j * J;
            const int cur_J     = std::min(col_start + J, T) - col_start;

            // Line 7: tile loads
            auto tt = Clock::now();
            MatrixXfRM Kt_j = Kt.middleCols(col_start, cur_J);
            MatrixXfRM V_j  = V.middleRows(col_start, cur_J);
            p.t7 += std::chrono::duration<double>(Clock::now() - tt).count();

            // Line 8: S = Q_i * Kt_j
            tt = Clock::now();
            MatrixXfRM S = Q_i * Kt_j;
            S *= scale;
            p.t8 += std::chrono::duration<double>(Clock::now() - tt).count();

            // Line 10: softmax statistics
            tt = Clock::now();
            Eigen::VectorXf m_new = m.cwiseMax(S.rowwise().maxCoeff());
            Eigen::VectorXf alpha = (m - m_new).array().exp();
            Eigen::MatrixXf P     = (S.colwise() - m_new).array().exp().matrix();
            Eigen::VectorXf l_new = alpha.cwiseProduct(l) + P.rowwise().sum();
            p.t10 += std::chrono::duration<double>(Clock::now() - tt).count();

            // Line 11: output accumulation
            tt = Clock::now();
            auto O_block = O.middleRows(row_start, cur_I);
            O_block.array().colwise() *= alpha.array();
            O.middleRows(row_start, cur_I) += P * V_j;
            p.t11 += std::chrono::duration<double>(Clock::now() - tt).count();

            m = m_new;
            l = l_new;
        }
        O.middleRows(row_start, cur_I).array().colwise() /= l.array();
    }

    p.runtime = std::chrono::duration<double>(Clock::now() - t0).count();
    p.others  = std::max(0.0, p.runtime - (p.t7 + p.t8 + p.t10 + p.t11));
    return {std::move(O), p};
}
